### PR TITLE
Pass the original input to all PP stages

### PR DESCRIPTION
**Pull request by fegin**
We need the original tokens to generate the document masks/block causal masks. Since TorchTitan currently let all ranks perform data loading, there will be no performance regressions.

This is required to support document masking attention with PP.

### CODE REVIEW

File: torchtitan/experiments/llama4/model/model.py [hunk 1]
```diff
@@ -441,7 +441,7 @@ def _precompute_freqs_cis(self) -> torch.Tensor:
             self.model_args.rope_theta,
         )
 
-    def forward(self, tokens: torch.Tensor):
+    def forward(self, tokens: torch.Tensor, input_batch: torch.Tensor):
         """
         Perform a forward pass through the Transformer model.
```
Comment (H-Huang) [inline torchtitan/experiments/llama4/model/model.py:447]: Update docstring "Args:" below

File: torchtitan/models/llama3/model.py [hunk 2]
```diff
@@ -462,7 +462,7 @@ def _precompute_freqs_cis(self) -> torch.Tensor:
             self.model_args.rope_theta,
         )
 
-    def forward(self, tokens: torch.Tensor):
+    def forward(self, tokens: torch.Tensor, input_batch: torch.Tensor):
         """
         Perform a forward pass through the Transformer model.
```
Comment (H-Huang) [inline torchtitan/models/llama3/model.py:469]: Update docstring "Args:" below

**Review by H-Huang — APPROVED** [review]
on PP side looks good. Does the current llama3 and 4 not support document masking? How come model changes are needed?

**Review by wconstab — APPROVED** [review]

File: torchtitan/experiments/llama4/model/model.py [hunk 3]
```diff
@@ -441,21 +441,24 @@ def _precompute_freqs_cis(self) -> torch.Tensor:
...
 
         Args:
-            tokens (torch.Tensor): Input token indices.
+            tokens (torch.Tensor): Input token indices if pipeline parallelism is not enabled.
+                If pipeline parallelism is enabled, this will be the input token indices
+                for the ranks on the first pipeline stage. This will be the activation of the
+                previous pipeline stage if the current rank is not on the first stage.
+            input_batch (torch.Tensor): The input batch read from the dataloader.
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/model/model.py:453]: please add a comment that -- this field is needed for non-first PP stages to obtain proper document masks

File: torchtitan/train.py [hunk 4]
```diff
@@ -351,7 +355,7 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
             # Non-PP forward / backward
             with self.train_context(optional_context_parallel_ctx):
                 assert len(model_parts) == 1
-                pred = model_parts[0](inputs)
+                pred = model_parts[0](inputs, input_batch=inputs)
```
Comment (tianyu-l) [inline torchtitan/train.py]: non PP branch looks a bit strange -- I slightly prefer the alternative way of making `input_batch` optional, and let `init_attention_mask` use `tokens` if `input_batch is None`. The idea is that non PP users see less universal usage of `input_batch`

**Review by tianyu-l — APPROVED** [review]
Left some nit comments.

CI is broken by compile + SAC. Please make sure this change works before merge.

File: torchtitan/train.py (hunk 4 above)
Comment (fegin) [reply torchtitan/train.py]: I'm a little bit concern that we are doing branches due to "parallelism" selection in the model code. `input_batch is None` is True if PP is not applied. This seems to be making the model code  be aware of the parallelism. But I guess using `input_batch` directly already exposes some parallelism information, lol.

Comment (tianyu-l) [reply torchtitan/train.py]: Yeah that's what I mean, instead of branching internally on a local piece. Here we are branching at a much higher level. But I don't have too strong opinion, you can make the call.

In the long term, I feel we might revisit the `eos_id` way of passing things around. E.g. with multimodal we need more masks, so maybe an class/object based on sequence info is more convenient to pass around.
